File: agent_matchmaking/federation.py
"""Federation router — adapter pattern for cross-platform registry queries.

Dispatches queries to multiple registries in parallel with configurable
timeout. Each registry is accessed through a RegistryAdapter subclass
(Section 7.1).
"""

import time as _time
from concurrent.futures import ThreadPoolExecutor, TimeoutError as FuturesTimeout
from typing import Any, Callable, Dict, List, Optional

from .schema import (
    FederationQuery,
    FederationResult,
    MatchRequest,
    UnifiedCapabilityProfile,
    _now_iso,
)
from .discovery import deduplicate, normalize_results, translate_to_federation_query


# ---------------------------------------------------------------------------
# Registry Adapter (abstract base)
# ---------------------------------------------------------------------------

class RegistryAdapter:
    """Base class for registry adapters (Section 7.1).

    Subclass and implement `search()` to connect a new registry.
    """

    name: str = "base"

    def search(self, query: FederationQuery) -> List[UnifiedCapabilityProfile]:
        """Execute a search against this registry. Override in subclasses."""
        raise NotImplementedError

    def health_check(self) -> bool:
        """Return True if the registry is reachable."""
        return True


class LocalStoreAdapter(RegistryAdapter):
    """Adapter that searches the local MatchmakingStore."""

    name = "local"

    def __init__(self, store: Any) -> None:
        self._store = store

    def search(self, query: FederationQuery) -> List[UnifiedCapabilityProfile]:
        return self._store.search_ucps(
            domain=query.domain,
            subdomain=query.subdomain,
            text=query.query_text,
        )


class StaticAdapter(RegistryAdapter):
    """Adapter backed by a static list of UCPs (useful for testing)."""

    name = "static"

    def __init__(self, ucps: List[UnifiedCapabilityProfile], name: str = "static") -> None:
        self._ucps = ucps
        self.name = name

    def search(self, query: FederationQuery) -> List[UnifiedCapabilityProfile]:
        text_lower = query.query_text.lower()
        results = []
        for ucp in self._ucps:
            for cap in ucp.capabilities:
                if (
                    (not query.domain or cap.domain == query.domain)
                    and (not text_lower or text_lower in cap.description.lower())
                ):
                    results.append(ucp)
                    break
        return results[:query.max_results]


class CallbackAdapter(RegistryAdapter):
    """Adapter that delegates to a callback function."""

    def __init__(
        self,
        name: str,
        search_fn: Callable[[FederationQuery], List[UnifiedCapabilityProfile]],
    ) -> None:
        self.name = name
        self._search_fn = search_fn

    def search(self, query: FederationQuery) -> List[UnifiedCapabilityProfile]:
        return self._search_fn(query)


# ---------------------------------------------------------------------------
# Federation Router
# ---------------------------------------------------------------------------

class FederationRouter:
    """Routes queries to registered adapters in parallel (Section 7.1).

    Architecture:
        Match Engine  <->  FederationRouter  <->  RegistryAdapter[]
    """

    def __init__(self, timeout_ms: int = 5000, max_workers: int = 10) -> None:
        self._adapters: Dict[str, RegistryAdapter] = {}
        self.timeout_ms = timeout_ms
        self.max_workers = max_workers

    def register(self, adapter: RegistryAdapter) -> None:
        self._adapters[adapter.name] = adapter

    def unregister(self, name: str) -> None:
        self._adapters.pop(name, None)

    @property
    def registry_names(self) -> List[str]:
        return list(self._adapters.keys())

    def query(
        self,
        request: MatchRequest,
        registries: Optional[List[str]] = None,
    ) -> List[FederationResult]:
        """Dispatch federation query to all (or specified) adapters in parallel.

        Returns one FederationResult per adapter.
        """
        fed_query = translate_to_federation_query(request)

        # Determine which adapters to query
        if registries and "all" not in registries:
            adapters = {n: a for n, a in self._adapters.items() if n in registries}
        else:
            adapters = dict(self._adapters)

        if not adapters:
            return []

        timeout_sec = self.timeout_ms / 1000.0
        results: List[FederationResult] = []

        def _query_adapter(name_adapter):
            name, adapter = name_adapter
            start = _time.monotonic()
            try:
                ucps = adapter.search(fed_query)
                elapsed = int((_time.monotonic() - start) * 1000)
                normalized = normalize_results(ucps, source_registry=name)
                return FederationResult(
                    registry_name=name,
                    ucps=normalized,
                    query_time_ms=elapsed,
                )
            except Exception as e:
                elapsed = int((_time.monotonic() - start) * 1000)
                return FederationResult(
                    registry_name=name,
                    ucps=[],
                    query_time_ms=elapsed,
                    error=str(e),
                )

        # Parallel dispatch (Step 2)
        workers = min(self.max_workers, len(adapters))
        with ThreadPoolExecutor(max_workers=workers) as pool:
            futures = {
                pool.submit(_query_adapter, (name, adapter)): name
                for name, adapter in adapters.items()
            }
            for future in futures:
                try:
                    result = future.result(timeout=timeout_sec)
                    results.append(result)
                except FuturesTimeout:
                    name = futures[future]
                    results.append(FederationResult(
                        registry_name=name,
                        ucps=[],
                        error="timeout",
                    ))
                except Exception as e:
                    name = futures[future]
                    results.append(FederationResult(
                        registry_name=name,
                        ucps=[],
                        error=str(e),
                    ))

        return results

    def federated_search(
        self,
        request: MatchRequest,
        registries: Optional[List[str]] = None,
    ) -> List[UnifiedCapabilityProfile]:
        """Full federated search: query all registries, normalize, deduplicate.

        Convenience method combining query + deduplication.
        """
        results = self.query(request, registries)

        all_ucps: List[UnifiedCapabilityProfile] = []
        for result in results:
            all_ucps.extend(result.ucps)

        return deduplicate(all_ucps)

File: typescript/src/federation.ts
import {
  FederationQuery,
  FederationResult,
  MatchRequest,
  UnifiedCapabilityProfile,
} from "./schema";
import { deduplicate, normalizeResults, translateToFederationQuery } from "./discovery";

// ---------------------------------------------------------------------------
// Registry Adapter (abstract base)
// ---------------------------------------------------------------------------

export abstract class RegistryAdapter {
  abstract readonly name: string;
  abstract search(query: FederationQuery): UnifiedCapabilityProfile[];
  healthCheck(): boolean { return true; }
}

export class LocalStoreAdapter extends RegistryAdapter {
  readonly name = "local";
  private store: { searchUcps(domain: string, subdomain: string, text: string): UnifiedCapabilityProfile[] };

  constructor(store: { searchUcps(domain: string, subdomain: string, text: string): UnifiedCapabilityProfile[] }) {
    super();
    this.store = store;
  }

  search(query: FederationQuery): UnifiedCapabilityProfile[] {
    return this.store.searchUcps(query.domain, query.subdomain, query.queryText);
  }
}

export class StaticAdapter extends RegistryAdapter {
  name: string;
  private ucps: UnifiedCapabilityProfile[];

  constructor(ucps: UnifiedCapabilityProfile[], name = "static") {
    super();
    this.ucps = ucps;
    this.name = name;
  }

  search(query: FederationQuery): UnifiedCapabilityProfile[] {
    const textLower = query.queryText.toLowerCase();
    const results: UnifiedCapabilityProfile[] = [];
    for (const ucp of this.ucps) {
      for (const cap of ucp.capabilities) {
        if (
          (!query.domain || cap.domain === query.domain) &&
          (!textLower || cap.description.toLowerCase().includes(textLower))
        ) {
          results.push(ucp);
          break;
        }
      }
    }
    return results.slice(0, query.maxResults);
  }
}

export class CallbackAdapter extends RegistryAdapter {
  name: string;
  private searchFn: (query: FederationQuery) => UnifiedCapabilityProfile[];

  constructor(name: string, searchFn: (query: FederationQuery) => UnifiedCapabilityProfile[]) {
    super();
    this.name = name;
    this.searchFn = searchFn;
  }

  search(query: FederationQuery): UnifiedCapabilityProfile[] {
    return this.searchFn(query);
  }
}

// ---------------------------------------------------------------------------
// Federation Router
// ---------------------------------------------------------------------------

export class FederationRouter {
  private adapters: Record<string, RegistryAdapter> = {};
  timeoutMs: number;
  maxWorkers: number;

  constructor(timeoutMs = 5000, maxWorkers = 10) {
    this.timeoutMs = timeoutMs;
    this.maxWorkers = maxWorkers;
  }

  register(adapter: RegistryAdapter): void {
    this.adapters[adapter.name] = adapter;
  }

  unregister(name: string): void {
    delete this.adapters[name];
  }

  get registryNames(): string[] {
    return Object.keys(this.adapters);
  }

  query(request: MatchRequest, registries?: string[]): FederationResult[] {
    const fedQuery = translateToFederationQuery(request);

    let adaptersToQuery: Record<string, RegistryAdapter>;
    if (registries && !registries.includes("all")) {
      adaptersToQuery = {};
      for (const [n, a] of Object.entries(this.adapters)) {
        if (registries.includes(n)) adaptersToQuery[n] = a;
      }
    } else {
      adaptersToQuery = { ...this.adapters };
    }

    if (Object.keys(adaptersToQuery).length === 0) return [];

    const results: FederationResult[] = [];

    for (const [name, adapter] of Object.entries(adaptersToQuery)) {
      const start = performance.now();
      try {
        const ucps = adapter.search(fedQuery);
        const elapsedMs = Math.round(performance.now() - start);
        const normalized = normalizeResults(ucps, name);
        results.push(new FederationResult({
          registryName: name,
          ucps: normalized,
          queryTimeMs: elapsedMs,
        }));
      } catch (e) {
        const elapsedMs = Math.round(performance.now() - start);
        results.push(new FederationResult({
          registryName: name,
          ucps: [],
          queryTimeMs: elapsedMs,
          error: String(e),
        }));
      }
    }

    return results;
  }

  federatedSearch(request: MatchRequest, registries?: string[]): UnifiedCapabilityProfile[] {
    const results = this.query(request, registries);
    const allUcps: UnifiedCapabilityProfile[] = [];
    for (const result of results) {
      allUcps.push(...result.ucps);
    }
    return deduplicate(allUcps);
  }
}
